R/cal-estimate.R

```diff
@@ -0,0 +1,405 @@
+                                       ...) {
+  UseMethod("cal_estimate_isotonic_boot")
+}
+
+#' @export
+cal_estimate_isotonic_boot.data.frame <- function(.data,
+                                                  truth = NULL,
+                                                  estimate = dplyr::starts_with(".pred_"),
+                                                  times = 10,
+                                                  ...) {
+  truth <- enquo(truth)
+
+  levels <- truth_estimate_map(.data, !!truth, {{ estimate }})
+
+  if (length(levels) == 2) {
+    log_model <- cal_isoreg_boot(
+      .data = .data,
+      truth = !!truth,
+      estimate = levels[[1]],
+      times = times,
+      ...
+    )
+
+    res <- as_binary_cal_object(
+      estimate = log_model,
+      levels = levels,
+      truth = !!truth,
+      method = "Isotonic Bootstrapped",
+      additional_class = "cal_estimate_isotonic_boot"
+    )
+  } else {
+    stop_multiclass()
+  }
+
+  res
+}
+
+#------------------------- Estimate implementation -----------------------------
+#------------------------------ >> Logistic ------------------------------------
+cal_estimate_logistic_impl <- function(.data,
+                                       truth = NULL,
+                                       estimate = dplyr::starts_with(".pred_"),
+                                       type,
+                                       model,
+                                       method,
+                                       additional_class,
+                                       ...) {
+  truth <- enquo(truth)
+
+  levels <- truth_estimate_map(.data, !!truth, {{ estimate }})
+
+  if (length(levels) == 2) {
+    log_model <- cal_model_impl(
+      .data = .data,
+      truth = !!truth,
+      estimate = levels[[1]],
+      run_model = model,
+      ...
+    )
+
+    res <- as_binary_cal_object(
+      estimate = log_model,
+      levels = levels,
+      truth = !!truth,
+      method = method,
+      additional_class = additional_class
+    )
+  } else {
+    stop_multiclass()
+  }
+
+  res
+}
+
+cal_model_impl <- function(.data, truth, estimate, run_model, ...) {
+  truth <- ensym(truth)
+
+  if (run_model == "logistic_spline") {
+    f_model <- expr(!!truth ~ s(!!estimate, k = 10))
+    init_model <- mgcv::gam(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  if (run_model == "glm") {
+    f_model <- expr(!!truth ~ !!estimate)
+    init_model <- glm(f_model, data = .data, family = "binomial", ...)
+    model <- butcher::butcher(init_model)
+  }
+
+  model
+}
+
+#------------------------------ >> Isotonic ------------------------------------
+cal_isoreg_dataframe <- function(.data,
+                                 truth,
+                                 estimate,
+                                 sampled = FALSE,
+                                 ...) {
+  sort_data <- dplyr::arrange(.data, !!estimate)
```

Comment:
Maybe call this `training_data` or something. `sort_data` is confusing.

Reply:
But this is after sorting the data via `arrange()`, figured it was the most appropriate name. Does it matter for an intermediate variable?

Reply:
`sorted_data`